diff --git a/crates/c_api/CMakeLists.txt b/crates/c_api/CMakeLists.txt
index b15c787..97dd294 100644
--- a/crates/c_api/CMakeLists.txt
+++ b/crates/c_api/CMakeLists.txt
@@ -43,6 +43,10 @@ endif()
 list(TRANSFORM WASMI_SHARED_FILES PREPEND ${WASMI_TARGET_DIR}/)
 list(TRANSFORM WASMI_STATIC_FILES PREPEND ${WASMI_TARGET_DIR}/)
 
+if(NOT BUILD_SHARED_LIBS)
+  set(WASMI_SHARED_FILES )
+endif()
+
 # Instructions on how to build and install the Wasmi Rust crate.
 find_program(WASMI_CARGO_BINARY cargo REQUIRED)
 include(ExternalProject)
