--- /usr/include/boost/config/compiler/clang.hpp 2013-07-20 13:17:10.000000000 -0400
+++ /usr/include/boost/config/compiler/clang.rippled.hpp 2014-03-11 16:40:51.000000000 -0400
@@ -39,6 +39,10 @@
 // Clang supports "long long" in all compilation modes.
 #define BOOST_HAS_LONG_LONG

+#if defined(__SIZEOF_INT128__)
+#  define BOOST_HAS_INT128
+#endif
+
 //
 // Dynamic shared object (DSO) and dynamic-link library (DLL) support
 //
