diff --git a/CMakeLists.txt b/CMakeLists.txt
index 93b884d..b715cb6 100644
--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -106,14 +106,9 @@ endif()
 include(CMakeDependentOption)

 if(MSVC)
-  option(WITH_GFLAGS "build with GFlags" OFF)
   option(WITH_XPRESS "build with windows built in compression" OFF)
-  option(ROCKSDB_SKIP_THIRDPARTY "skip thirdparty.inc" OFF)
-
-  if(NOT ROCKSDB_SKIP_THIRDPARTY)
-    include(${CMAKE_CURRENT_SOURCE_DIR}/thirdparty.inc)
-  endif()
-else()
+endif()
+if(TRUE)
   if(CMAKE_SYSTEM_NAME MATCHES "FreeBSD" AND NOT CMAKE_SYSTEM_NAME MATCHES "kFreeBSD")
     # FreeBSD has jemalloc as default malloc
     # but it does not have all the jemalloc files in include/...
@@ -126,7 +121,7 @@ else()
     endif()
   endif()

-  if(MINGW)
+  if(MSVC OR MINGW)
     option(WITH_GFLAGS "build with GFlags" OFF)
   else()
     option(WITH_GFLAGS "build with GFlags" ON)
