// ripple-lib/dist/npm/api/index.js


'use strict';
var _ = require('lodash');
var common = require('./common');
var server = require('./server/server');
var connect = server.connect;
var disconnect = server.disconnect;
var getServerInfo = server.getServerInfo;
var getFee = server.getFee;
var isConnected = server.isConnected;
var getLedgerVersion = server.getLedgerVersion;
var getTransaction = require('./ledger/transaction');
var getTransactions = require('./ledger/transactions');
var getTrustlines = require('./ledger/trustlines');
var getBalances = require('./ledger/balances');
var getPaths = require('./ledger/pathfind');
var getOrders = require('./ledger/orders');
var getOrderbook = require('./ledger/orderbook');
var getSettings = require('./ledger/settings');
var getAccountInfo = require('./ledger/accountinfo');
var preparePayment = require('./transaction/payment');
var prepareTrustline = require('./transaction/trustline');
var prepareOrder = require('./transaction/order');
var prepareOrderCancellation = require('./transaction/ordercancellation');
var prepareSettings = require('./transaction/settings');
var sign = require('./transaction/sign');
var submit = require('./transaction/submit');
var errors = require('./common').errors;
var convertExceptions = require('./common').convertExceptions;
var generateAddress = convertExceptions(common.generateAddress);
var computeLedgerHash = require('./offline/ledgerhash');
var getLedger = require('./ledger/ledger');
var isValidAddress = common.isValidAddress;

function RippleAPI(options) {
  common.validate.remoteOptions(options);
  var _options = _.assign({}, options, { automatic_resubmission: false });
  this.remote = new common.core.Remote(_options);
}

RippleAPI.prototype = {
  connect: connect,
  disconnect: disconnect,
  isConnected: isConnected,
  getServerInfo: getServerInfo,
  getFee: getFee,
  getLedgerVersion: getLedgerVersion,

  getTransaction: getTransaction,
  getTransactions: getTransactions,
  getTrustlines: getTrustlines,
  getBalances: getBalances,
  getPaths: getPaths,
  getOrders: getOrders,
  getOrderbook: getOrderbook,
  getSettings: getSettings,
  getAccountInfo: getAccountInfo,
  getLedger: getLedger,

  preparePayment: preparePayment,
  prepareTrustline: prepareTrustline,
  prepareOrder: prepareOrder,
  prepareOrderCancellation: prepareOrderCancellation,
  prepareSettings: prepareSettings,
  sign: sign,
  submit: submit,

  computeLedgerHash: computeLedgerHash,
  isValidAddress: isValidAddress,
  generateAddress: generateAddress,
  errors: errors
};

// these are exposed only for use by unit tests; they are not part of the API
RippleAPI._PRIVATE = {
  common: common,
  ledgerUtils: require('./ledger/utils'),
  schemaValidator: require('./common/schema-validator')
};

module.exports = RippleAPI;
